File: problem12/vers_java/P12.java
/*
 * P12.java
 *
 * Zeke Reyna
 *
 * Finds the first triangular number with more than
 * the given number of divisors
 */

import static java.lang.System.out;
public class P12
{
        public static int find_tri( int x )
        {
                return (x * x + x) / 2;
        }

        public static int find_num_factors( int num )
        {
                int i;
                int num_factors, power;

                num_factors = 1;
                for (i = 2; i < (int) Math.sqrt(num); i++) {
                        power = 0;

                        while (num % i == 0) {
                                num /= i;
                                power++;
                        }
                        num_factors *= (power + 1);
                }

                if (num > 1)
                        num_factors *= 2;

                return num_factors;
        }

        public static void main(String args[] )
        {
                int i, tri;
                int divis, curr_num_divis;

                if( args.length == 1 )
                        divis = Integer.parseInt( args[0] );
                else
                        divis = 5; // default

                for( i = 1; ; i++ ) {
                        tri = find_tri( i );

                        curr_num_divis = find_num_factors( tri );
                        if( curr_num_divis > divis )
                                break;
                }
                // do printing stuff and make func find_num_factors
                out.printf("\nThe first triangular number that has greater");
                out.printf(" than %d divisors is %d\n\n", divis, tri );
        }
}

File: problem12/vers_py/p12.py
### p12.py
#
# Zeke Reyna
#
# Finds the first triangular number
# that has more than the given amount
# of divisors

import sys
import math

def find_tri(x):
    return int((x * x + x) / 2)

def find_num_factors(n):

    num_factors = 1
    for i in range(2, int(math.sqrt(n))):
        power = 0

        while n % i == 0:
            n /= i
            power += 1
        num_factors *= (power + 1)

    if n > 1:
        num_factors *= 2

    return num_factors


def main(argv):

    if len(argv) == 2:
        divis = int(argv[1])
    else:
        divis = 5

    i = 1
    found = False
    while not found:
        tri = find_tri(i)
        i += 1

        curr_num_divis = find_num_factors(tri)
        if curr_num_divis > divis:
            found = True

    print ("\nThe first triangular number that has",
           " greater than ",divis," divisors is ",
           tri, "\n" )

main(sys.argv)
